server: introduce a serial console backend

gjcolombo opened the pull request:
Rework parts of `propolis-server`'s serial console logic to achieve the following goals:

- Make the serial console history buffer (which is transferred in the live migration protocol) use the versioned payload helpers supplied by the main Propolis library.
- Provide the primitives necessary to implement the policies described in RFD 491: it should be possible to distinguish read-write and read-only clients of a serial console and apply different policies (around the number of clients and what happens if they fall behind when reading) to each of them.
- Simplify the logic for dealing with multiple serial console clients to avoid cancellation safety issues.

To do this, split the existing serial console task into two layers. The `ConsoleBackend` layer deals with I/O to and from a Propolis character device. The `SerialConsoleManager` layer creates a task for each websocket connection to a serial device and registers it as a client of the relevant `ConsoleBackend`. The backend layer allows the manager to declare that specific clients have read-only access to the console and to decide what happens if a client becomes unable to accept new bytes from a client; the manager implements the policies that determine how many clients can exist concurrently and what backend policies apply to them.

Update the serial API endpoint to allow users to request a writable connection to the console. Add parameters to the serial console helper types in `propolis-client` that govern how this is set. This will require a small update to Nexus the next time Omicron picks up a Propolis build to add the extra parameter. End users don't see this aspect of the serial API (Nexus serves as a proxy for connections to VMs' serial consoles), so this should be a non-breaking change for control plane API users.

`propolis-cli serial` defaults to a writable connection to match previous behavior, but it now supports a `--readonly` flag to create read-only connections.

As currently written, this PR implements the "only one read-write client at a time" policy described in RFD 491. I can easily disable this part of the change if we don't want to do this yet (should just need to remove a few lines from the console manager's client-attach function).

Tests: cargo test, PHD, ad hoc testing with local propolis-server processes. This change changes the order of messages sent in the migration protocol and so will break the migrate-from-base tests; this remains OK for now since live migration isn't being used in production yet.

Fixes #650. Related to #614.

hawkw commented on bin/propolis-server/src/lib/serial/backend.rs:
```diff
@@ -0,0 +1,397 @@
+    Close,
+}
+
+/// An individual client of a character backend.
+struct Client {
+    /// Bytes read from the character device should be sent to the client on
+    /// this channel.
+    tx: mpsc::Sender<u8>,
+
+    /// Determines what happens when the backend wants to send a byte and
+    /// [`Self::tx`] is full.
+    read_discipline: FullReadChannelDiscipline,
+}
+
+/// A handle held by a client that represents its connection to the backend.
+pub(super) struct ClientHandle {
+    /// The client's ID.
+    id: ClientId,
+
+    /// A reference to the backend to which the client is connected.
+    backend: Arc<ConsoleBackend>,
+
+    /// The client's backend access rights.
+    permissions: Permissions,
+}
```
considered nitpicky, take it or leave it: personally, since the `ClientHandle` type represents the external API for consumers of this module, while the `Client` type is an internal-only implementation detail, I might've given the more consicse name to the interface type --- perhaps `Client` and `ClientState`, or something?

this isn't a big deal, though; if renaming it is too much of a pain, feel free to ignore me.

hawkw commented on bin/propolis-server/src/lib/serial/backend.rs:
```diff
@@ -0,0 +1,397 @@
+struct Client {
+    /// Bytes read from the character device should be sent to the client on
+    /// this channel.
+    tx: mpsc::Sender<u8>,
```
`mpsc::Sender<u8>` is a pretty unfortunate way to represent an asynchronous stream of bytes, from a performance standpoint: this will allocate a node in a linked list for each individual byte. If you want an asynchronous byte pipe, I would consider using something like [`tokio::io::simplex`](https://docs.rs/tokio/latest/tokio/io/fn.simplex.html) , which is an asynchronous byte pipe implemented as a single contiguous in memory buffer. There's also [`DuplexStream`](https://docs.rs/tokio/latest/tokio/io/struct.DuplexStream.html), which is the same thing but bi-directional.

gjcolombo replied:
`simplex` definitely seems like a much nicer tool here. The one gotcha with it is that `Sender` is `Clone`, but `SimplexStream` and its `WriteHalf` aren't, so the code in `read_task` that captures channels under the lock and then drops the lock before sending (so that it can guiltlessly block as needed) won't work as-is.

Now that I think about it, I wonder if it'd be better to rewrite `read_task` so that it doesn't need to look at the client map at all. (I think there's even an outside chance I could get rid of the client map altogether...) I'm going to give this a try to see how it shakes out.

hawkw commented on bin/propolis-server/src/lib/serial/backend.rs:
```diff
@@ -0,0 +1,397 @@
+    ///
+    /// - `Ok(bytes)` if the write succeeded; `bytes` is the number of bytes
+    ///   that were written.
+    /// - `Err(ErrorKind::PermissionDenied)` if this handle does not grant write
+    ///   access to the device.
+    /// - `Err(ErrorKind::ConnectionAborted)` if this handle's client was
```
nitpicky, considered unimportant: this doc comment is *technically* lying; this doesn't return an `ErrorKind`, but a whole `io::Error`...

hawkw commented on bin/propolis-server/src/lib/serial/backend.rs:
```diff
@@ -0,0 +1,397 @@
+            }
+
+            res = buf.read(bytes.as_mut_slice(), source.as_ref()) => {
+                res.unwrap()
```
nit: do we expect that this will never be an error? it might be nice to `expect` this with a string explaining why...

hawkw commented on bin/propolis-server/src/lib/serial/backend.rs:
```diff
@@ -0,0 +1,397 @@
+    /// A history of the bytes read from this backend's device.
+    buffer: HistoryBuffer,
+
+    /// A table mapping client IDs to clients.
+    clients: BTreeMap<ClientId, Client>,
+
+    /// The ID to assign to the next client to attach to this backend.
+    next_client_id: u64,
```
Rather than a `BTreeMap` and counter of IDs, I might consider using the [`slab`](https://crates.io/crates/slab) crate, which implements a collection of "things" indexed by integer ID, but stored in an array (so both iteration and lookups are likely much more efficient). Do note that `slab` will reuse indices of removed elements, in case that's an issue here, but AFAICT, it wouldn't actually be.

hawkw commented on bin/propolis-server/src/lib/serial/mod.rs:
```diff
@@ -2,400 +2,533 @@
+/// The collection of [`ClientTask`]s belonging to a console manager.
+#[derive(Default)]
+struct ClientTasks {
+    /// A table mapping task IDs to task structs.
+    tasks: BTreeMap<ClientId, ClientTask>,
```
similarly, i might consider whether this could be a `slab`.

hawkw commented on bin/propolis-server/src/lib/serial/mod.rs:
```diff
@@ -2,400 +2,533 @@
+            };
+
+            let task = ClientTask {
+                hdl: tokio::spawn(async move { serial_task(ctx).await }),
```
style nit: calling an async function, such as `serial_task`, returns a `Future`, so the `async move { ....await }` is unnecessary; this could just be: 
```suggestion
                hdl: tokio::spawn(serial_task(ctx)),
```

hawkw commented on bin/propolis-server/src/lib/serial/mod.rs:
```diff
@@ -2,400 +2,533 @@
-        tokio::select! {
-            // Poll in the order written
+        let event = select! {
+            // The priority of these branches is important:
+            //
+            // 1. Requests to stop the client take precedence over everything
+            //    else.
+            // 2. New bytes written by the guest need to be processed before any
+            //    other requests: if a guest outputs a byte while a read-write
+            //    client is attached, the relevant vCPU will be blocked until
+            //    the client processes the byte.
```
thank you for documenting this <3

gjcolombo commented:
Thanks as always for the thorough review, @hawkw! I'm going to rebase this as-is onto #834 and then will start tightening things up.

gjcolombo commented:
@hawkw: I think I've addressed all your feedback in the newest round of commits. The new changes use `tokio::io::simplex` (thank you for suggesting this!), but this required some architectural adjustments to account for differences between a `WriteHalf` and an `mpsc::Sender` (the former isn't `Clone`, and writes won't return an error if the reader is dropped). For the most part, what this amounts to is that the `ConsoleBackend`'s read task no longer reads the backend's client table directly; instead the backend feeds it a set of `Reader` objects, and readers and their clients share a `watch` so they can mutually shut down without having to hold references to each other.

There's a new theory statement in `serial/backend.rs` that tries to explain this a bit more. Let me know what you think.
